python/examples/atomistic/3-atomistic-model-with-nl.py

```diff
@@ -0,0 +1,300 @@
+from metatensor.torch.atomistic import (
+    MetatensorAtomisticModel,
+    ModelCapabilities,
+    ModelMetadata,
+    ModelOutput,
+    NeighborListOptions,
+    System,
+)
+
+# Integration with ASE calculator for metatensor atomistic models
+from metatensor.torch.atomistic.ase_calculator import MetatensorCalculator
+
+
+# %%
+#
+# A Lennard-Jones Model
+# ---------------------
+#
+# The Lennard-Jones potential is a mathematical model to approximate the interaction
+# between a pair of neutral atoms or molecules. It is given by the equation:
+#
+# .. math:: V(r) = 4\epsilon \left[ \left( \frac{\sigma}{r} \right)^{12} - \left(
+#     \frac{\sigma}{r} \right)^{6} \right]
+#
+# where :math:`\epsilon` is the depth of the potential well, :math:`\sigma` is the
+# finite distance at which the inter-particle potential is zero, and :math:`r` is the
+# distance between the particles. The 12-6 form is chosen because the :math:`r^12` term
+# approximates the Pauli repulsion at short ranges, while the :math:`r^6` term
+# represents the attractive van der Waals forces. The :math:`r^{12}` is choosen becuase
+# it is just the square of the :math:`r^6` and therefore allows fast evaluation.
+#
+# A Lennard-Jones potential is well-suited for argon because it accurately represents
+# the van der Waals forces that dominate the interactions between argon atoms especially
+# as for all noble gases. This potential was used in one of the first MD simulations:
+# "Correlation in the motions of atoms in liquid Argon" by A. Rahman (`Phys. Rev. 136,
+# A405-A411, 1964 <https://link.aps.org/doi/10.1103/PhysRev.136.A405>`_), which
+# demonstrated the effectiveness of continuous potentials in MD.
+#
+# The model below is a simplified version of a `more complex Lennard-Jones model
+# <https://github.com/Luthaf/metatensor-lj-test/blob/main/src/metatensor_lj_test/pure.py>`_.
+# The linked version also implements ``per_atom`` energies as well as atom selection
+# using the ``selected_atoms`` parameter of the ``forward()`` method. In this model, we
+# shift the energy by its value at the ``cutoff``. This will break the conservativeness
+# of the potential, which is unproblematic in here because we use a large cutoff and
+# therefore the potential already almost decayed to zero. For more sophisticated methods
+# like a polynomial switching potential like XPLOR, we refer to the literature.
+
+
+class LennardJonesModel(torch.nn.Module):
+    """Implementation of a single particle type Lennard-Jones potential."""
+
+    def __init__(self, cutoff, sigma, epsilon):
+        super().__init__()
+        self._nl_options = NeighborListOptions(cutoff=cutoff, full_list=False)
+
+        self._sigma = sigma
+        self._epsilon = epsilon
+
+        # shift the energy to 0 at the cutoff
+        self._shift = 4 * epsilon * ((sigma / cutoff) ** 12 - (sigma / cutoff) ** 6)
+
+    def requested_neighbor_lists(self) -> List[NeighborListOptions]:
```

Comment:
I would add a bunch of docs around this function to explain what it is doing/how it is used